api/src/main/kotlin/com/google/devtools/ksp/processing/Resolver.kt

```diff
@@ -75,4 +75,55 @@ interface Resolver {
      * map a declaration to jvm signature.
      */
     fun mapToJvmSignature(declaration: KSDeclaration): String
+
+    /**
+     * Returns the type of the [property] when it is viewed as member of the [containing] type.
+     *
+     * For instance, for the following input:
+     * ```
+     * class Base<T>(val x:T)
+     * val foo: Base<Int>
+     * val bar: Base<String>
+     * ```
+     * When `x` is viewed as member of `foo`, this method will return the [KSType] for `Int`
+     * whereas when `x` is viewed as member of `bar`, this method will return the [KSType]
+     * representing `String`.
+     *
+     * If the given type does not contain the [property], the [KSPropertyDeclaration.type] of the
+     * [property] will be returned.
```

Comment:
Would an explicit `KSErrorType` be less surprising to users?

Reply:
The usage pattern we have in Room is that we don't really care whether it makes sense or not.
But also, we don't really call it in a case where Type may not be a container. I would guess it is not really a common case for processors but I don't have data to back that claim :).

If we want to detect errors, i think it is better to throw `IllegalArgumentException` instead of an Error type.

Reply:
`IllegalArgumentException` sounds good. I'm just worrying about that returning the default type might mute some problems which should be caught earlier.

Reply:
replace w/ illegal argument, added tests.